File: simplified-cardcreator/src/main/java/org/nypl/simplified/cardcreator/fragments/HomeAddressFragment.java
package org.nypl.simplified.cardcreator.fragments;


import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.text.Editable;
import android.text.TextUtils;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;
import android.widget.TextView;

import org.nypl.simplified.prefs.Prefs;
import org.nypl.simplified.cardcreator.R;
import org.nypl.simplified.cardcreator.listener.InputListenerType;


/**
 * A simple {@link Fragment} subclass.
 * Use the {@link HomeAddressFragment#newInstance} factory method to
 * create an instance of this fragment.
 */
public class HomeAddressFragment extends Fragment {


    private EditText line_1;
    private EditText line_2;
    private EditText city;
    private EditText state;
    private EditText zip;
    private Prefs prefs;

    /**
     * @return address line 1
     */
    public EditText getLine_1() {
        return this.line_1;
    }

    /**
     * @return address line 2
     */
    public EditText getLine_2() {
        return this.line_2;
    }

    /**
     * @return address city
     */
    public EditText getCity() {
        return this.city;
    }

    /**
     * @return address state
     */
    public EditText getState() {
        return this.state;
    }

    /**
     * @return address zip
     */
    public EditText getZip() {
        return this.zip;
    }

    /**
     *
     */
    public HomeAddressFragment() {
        // Required empty public constructor
    }

    /**
     * Use this factory method to create a new instance of
     * this fragment using the provided parameters.
     *
     * @return A new instance of fragment HomeAddressFragment.
     */

    public  HomeAddressFragment newInstance() {
        final HomeAddressFragment fragment = new HomeAddressFragment();
        final Bundle args = new Bundle();

        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(final Bundle instance_state) {
        super.onCreate(instance_state);

        this.prefs = new Prefs(getContext());
    }

    @Override
    public View onCreateView(final LayoutInflater inflater,
                             final ViewGroup container,
                             final Bundle instance_state) {
        // Inflate the layout for this fragment
        final View root_view = inflater.inflate(R.layout.fragment_home_address, container, false);
        ((TextView) root_view.findViewById(android.R.id.title)).setText("Home Address");

        this.line_1 = ((EditText) root_view.findViewById(R.id.street1));
        this.line_2 = ((EditText) root_view.findViewById(R.id.street2));
        this.city = ((EditText) root_view.findViewById(R.id.city));
        this.state = ((EditText) root_view.findViewById(R.id.region));
        this.zip = ((EditText) root_view.findViewById(R.id.zip));


        this.line_1.setText(this.prefs.getString(getResources().getString(R.string.STREET1_H_DATA_KEY)));
        this.line_2.setText(this.prefs.getString(getResources().getString(R.string.STREET2_H_DATA_KEY)));
        this.city.setText(this.prefs.getString(getResources().getString(R.string.CITY_H_DATA_KEY)));
        this.state.setText(this.prefs.getString(getResources().getString(R.string.STATE_H_DATA_KEY)));
        this.zip.setText(this.prefs.getString(getResources().getString(R.string.ZIP_H_DATA_KEY)));

        if (this.isCompleted()) {
            ((InputListenerType) getActivity()).onInputComplete();
        }
        else {
            ((InputListenerType) getActivity()).onInputInComplete();
        }


        this.line_1.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(final CharSequence s, final int start, final int count, final int after) {
            }

            @Override
            public void onTextChanged(final CharSequence s, final int start, final int before, final int count) {
            }

            @Override
            public void afterTextChanged(final Editable s) {
                HomeAddressFragment.this.prefs.putString(getResources().getString(R.string.STREET1_H_DATA_KEY), (s != null) ? s.toString() : null);

                if (HomeAddressFragment.this.isCompleted()) {
                    ((InputListenerType) getActivity()).onInputComplete();
                }
                else {
                    ((InputListenerType) getActivity()).onInputInComplete();
                }
            }
        });
        this.line_2.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(final CharSequence s, final int start, final int count, final int after) {
            }

            @Override
            public void onTextChanged(final CharSequence s, final int start, final int before, final int count) {
            }

            @Override
            public void afterTextChanged(final Editable s) {
                HomeAddressFragment.this.prefs.putString(getResources().getString(R.string.STREET2_H_DATA_KEY), (s != null) ? s.toString() : null);

                if (HomeAddressFragment.this.isCompleted()) {
                    ((InputListenerType) getActivity()).onInputComplete();
                }
                else {
                    ((InputListenerType) getActivity()).onInputInComplete();
                }
            }
        });
        this.city.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(final CharSequence s, final int start, final int count, final int after) {
            }

            @Override
            public void onTextChanged(final CharSequence s, final int start, final int before, final int count) {
            }

            @Override
            public void afterTextChanged(final Editable s) {
                HomeAddressFragment.this.prefs.putString(getResources().getString(R.string.CITY_H_DATA_KEY), (s != null) ? s.toString() : null);

                if (HomeAddressFragment.this.isCompleted()) {
                    ((InputListenerType) getActivity()).onInputComplete();
                }
                else {
                    ((InputListenerType) getActivity()).onInputInComplete();
                }
            }
        });
        this.state.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(final CharSequence s, final int start, final int count, final int after) {
            }

            @Override
            public void onTextChanged(final CharSequence s, final int start, final int before, final int count) {
            }

            @Override
            public void afterTextChanged(final Editable s) {
                HomeAddressFragment.this.prefs.putString(getResources().getString(R.string.STATE_H_DATA_KEY), (s != null) ? s.toString() : null);

                if (HomeAddressFragment.this.isCompleted()) {
                    ((InputListenerType) getActivity()).onInputComplete();
                }
                else {
                    ((InputListenerType) getActivity()).onInputInComplete();
                }
            }
        });
        this.zip.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(final CharSequence s, final int start, final int count, final int after) {
            }

            @Override
            public void onTextChanged(final CharSequence s, final int start, final int before, final int count) {
            }

            @Override
            public void afterTextChanged(final Editable s) {
                HomeAddressFragment.this.prefs.putString(getResources().getString(R.string.ZIP_H_DATA_KEY), (s != null) ? s.toString() : null);

                if (HomeAddressFragment.this.isCompleted()) {
                    ((InputListenerType) getActivity()).onInputComplete();
                }
                else {
                    ((InputListenerType) getActivity()).onInputInComplete();
                }
            }
        });
        return root_view;
    }

    /**
     * @return boolean all required field completed
     */
    public boolean isCompleted() {
        return !TextUtils.isEmpty(this.prefs.getString(getResources().getString(R.string.STREET1_H_DATA_KEY)))
                && !TextUtils.isEmpty(this.prefs.getString(getResources().getString(R.string.CITY_H_DATA_KEY)))
                && !TextUtils.isEmpty(this.prefs.getString(getResources().getString(R.string.STATE_H_DATA_KEY)))
                && !TextUtils.isEmpty(this.prefs.getString(getResources().getString(R.string.ZIP_H_DATA_KEY)));
    }


}

File: simplified-cardcreator/src/main/java/org/nypl/simplified/cardcreator/ui/HomeAddressFragment.kt
package org.nypl.simplified.cardcreator.ui

import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView
import android.widget.ArrayAdapter
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.fragment.app.Fragment
import androidx.fragment.app.activityViewModels
import androidx.navigation.NavController
import androidx.navigation.NavDirections
import androidx.navigation.Navigation
import org.nypl.simplified.cardcreator.R
import org.nypl.simplified.cardcreator.databinding.FragmentHomeAddressBinding
import org.nypl.simplified.cardcreator.model.Address
import org.nypl.simplified.cardcreator.model.ValidateAddressResponse
import org.nypl.simplified.cardcreator.utils.Cache
import org.nypl.simplified.cardcreator.utils.hideKeyboard
import org.nypl.simplified.cardcreator.viewmodel.AddressViewModel
import org.slf4j.LoggerFactory

class HomeAddressFragment : Fragment(), AdapterView.OnItemSelectedListener {

  private val logger = LoggerFactory.getLogger(HomeAddressFragment::class.java)

  private var _binding: FragmentHomeAddressBinding? = null
  private val binding get() = _binding!!

  private lateinit var navController: NavController
  private lateinit var nextAction: NavDirections
  private val addressCharsMin = 5
  private var dialog: AlertDialog? = null

  private val viewModel: AddressViewModel by activityViewModels()

  override fun onCreateView(
    inflater: LayoutInflater,
    container: ViewGroup?,
    savedInstanceState: Bundle?
  ): View {
    _binding = FragmentHomeAddressBinding.inflate(inflater, container, false)
    return binding.root
  }

  override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
    super.onViewCreated(view, savedInstanceState)

    navController = Navigation.findNavController(requireActivity(), R.id.card_creator_nav_host_fragment)

    binding.spState.onItemSelectedListener = this
    ArrayAdapter.createFromResource(
      requireContext(),
      R.array.states_array,
      android.R.layout.simple_spinner_item
    ).also { adapter ->
      adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
      binding.spState.adapter = adapter
    }

    binding.etCity.addTextChangedListener(object : TextWatcher {
      override fun afterTextChanged(s: Editable?) {
        validateForm()
      }

      override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {}
      override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {}
    })

    binding.etStreet1.addTextChangedListener(object : TextWatcher {
      override fun afterTextChanged(s: Editable?) {
        validateForm()
      }

      override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {}
      override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {}
    })

    binding.etZip.addTextChangedListener(object : TextWatcher {
      override fun afterTextChanged(s: Editable?) {
        validateForm()
      }

      override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {}
      override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {}
    })

    // Go to next screen
    binding.nextBtn.setOnClickListener {
      validateAddress()
    }

    // Go to previous screen
    binding.prevBtn.setOnClickListener {
      navController.popBackStack()
    }

    viewModel.validateAddressResponse
      .receive(viewLifecycleOwner, this::handleValidateAddressResponse)

    viewModel.pendingRequest.observe(viewLifecycleOwner, this::showLoading)

    restoreViewData()
  }

  /**
   * Validates entered address
   */
  private fun validateAddress() {
    viewModel.validateAddress(
      Address(
        line1 = binding.etStreet1.text.toString(),
        city = binding.etCity.text.toString(),
        state = getStateAbbreviation(binding.spState.selectedItem.toString()),
        zip = binding.etZip.text.toString(),
        isResidential = true,
        hasBeenValidated = false
      )
    )
  }

  private fun handleValidateAddressResponse(response: ValidateAddressResponse) {
    when (response) {
      is ValidateAddressResponse.ValidateAddressData -> {
        logger.debug("Address is valid")
        goNextIfResidential(response.address)
      }
      is ValidateAddressResponse.AlternateAddressesError -> {
        logger.debug("Using first alternate address")
        goNextIfResidential(response.addresses.first())
      }
      is ValidateAddressResponse.ValidateAddressError -> {
        if (response.isUnrecognizedAddress) {
          val message = getString(R.string.unrecognized_address)
          Toast.makeText(activity, message, Toast.LENGTH_SHORT).show()
        } else {
          val message = getString(R.string.validate_address_error, response.status)
          showTryAgainDialog(message)
        }
      }
      is ValidateAddressResponse.ValidateAddressException -> {
        val message = getString(R.string.validate_address_general_error)
        showTryAgainDialog(message)
      }
    }
  }

  private fun goNextIfResidential(address: Address) {
    if (address.isResidential) {
      logger.debug("Address is residential")
      Cache(requireContext()).setHomeAddress(address)
      nextAction = HomeAddressFragmentDirections.actionNext()
      navController.navigate(nextAction)
    } else {
      logger.debug("Address is not residential")
      val message = getString(R.string.non_residential_address)
      Toast.makeText(activity, message, Toast.LENGTH_SHORT).show()
    }
  }

  private fun showTryAgainDialog(message: String) {
    val dialogBuilder = AlertDialog.Builder(requireContext())
    dialogBuilder.setMessage(message)
      .setCancelable(false)
      .setPositiveButton(getString(R.string.try_again)) { _, _ ->
        validateAddress()
      }
      .setNegativeButton(getString(R.string.cancel)) { dialog, _ ->
        dialog.cancel()
      }
    if (dialog == null) {
      dialog = dialogBuilder.create()
    }
    dialog?.show()
  }

  /**
   * Show loading screen
   */
  private fun showLoading(show: Boolean) {
    hideKeyboard()
    logger.debug("Toggling loading screen")
    if (show) {
      binding.loading.visibility = View.VISIBLE
      binding.form.visibility = View.GONE
      binding.navButtons.visibility = View.GONE
    } else {
      binding.loading.visibility = View.GONE
      binding.form.visibility = View.VISIBLE
      binding.navButtons.visibility = View.VISIBLE
    }
  }

  /**
   * Gets the abbreviation portion from the address state spinner
   */
  private fun getStateAbbreviation(stateListItem: String): String {
    return stateListItem.takeLast(3).dropLast(1)
  }

  /**
   * Checks form to verify that that address entered is legit
   */
  private fun validateForm() {
    binding.nextBtn.isEnabled = (
      binding.spState.selectedItem.toString() != getString(R.string.required) &&
        binding.etZip.text.length == addressCharsMin &&
        binding.etStreet1.text.length >= addressCharsMin &&
        binding.etCity.text.length >= addressCharsMin
      )
  }

  override fun onNothingSelected(parent: AdapterView<*>?) {
    binding.nextBtn.isEnabled = false
  }

  override fun onItemSelected(parent: AdapterView<*>?, view: View?, position: Int, id: Long) {
    logger.debug(parent!!.getItemAtPosition(position).toString())
    validateForm()
  }

  override fun onDestroyView() {
    super.onDestroyView()
    _binding = null
  }

  /**
   * Restores cached data
   */
  private fun restoreViewData() {
    val homeAddress = Cache(requireContext()).getHomeAddress()
    binding.etZip.setText(homeAddress.zip, TextView.BufferType.EDITABLE)
    binding.etStreet1.setText(homeAddress.line1, TextView.BufferType.EDITABLE)
    binding.etCity.setText(homeAddress.city, TextView.BufferType.EDITABLE)
    resources.getStringArray(R.array.states_array)
      .indexOfFirst { it.endsWith("(${homeAddress.state})") }.takeUnless { it == -1 }
      ?.let { binding.spState.setSelection(it) }
  }

  override fun onPause() {
    super.onPause()
    dialog?.dismiss()
  }
}
